code/controllers/subsystem/processing/overlays.dm

```diff
@@ -0,0 +1,115 @@
+var/datum/subsystem/processing/overlays/SSoverlays
+
+/datum/subsystem/processing/overlays
+	name = "Overlays"
+	flags = SS_TICKER|SS_NO_INIT|SS_FIRE_IN_LOBBY|SS_KEEP_TIMING
+	wait = 1
+	priority = 500
+	can_fire = FALSE
+
+	stat_tag = "Ov"
+	currentrun = null
+
+/datum/subsystem/processing/overlays/New()
+	NEW_SS_GLOBAL(SSoverlays)
+
+/datum/subsystem/processing/overlays/Recover()
+	processing = SSoverlays.processing
+
+/datum/subsystem/processing/overlays/fire()
+	while(processing.len)
+		var/atom/thing = processing[processing.len]
+		processing.len--
+		if(thing)
+			if(LAZYLEN(thing.priority_overlays) && LAZYLEN(thing.our_overlays))
+				thing.overlays = (thing.our_overlays | thing.priority_overlays)
+			else if(LAZYLEN(thing.our_overlays))
+				thing.overlays = thing.our_overlays.Copy()
+			else if(LAZYLEN(thing.priority_overlays))			//Do these two have to be copied? We don't want assignments to them to trigger overlay updates. But, idk, bruh, BYOND. - Cyberboss
+				thing.overlays = thing.priority_overlays.Copy()
+			else
+				LAZYCLEARLIST(thing.overlays)
+		if(MC_TICK_CHECK)
+			break
+	can_fire = processing.len > 0
+
+/atom/proc/cut_overlays(priority = FALSE)
+	var/list/cached_overlays = our_overlays
+	var/list/cached_priority = priority_overlays
+	
+	var/need_compile = FALSE
+
+	if(LAZYLEN(cached_overlays)) //don't queue empty lists, don't cut priority overlays
+		cached_overlays.Cut()  //clear regular overlays
+		need_compile = TRUE
+
+	if(priority && LAZYLEN(cached_priority))
+		cached_priority.Cut()
+		need_compile = TRUE
+
+	if(need_compile)
+		SSoverlays.processing[src] = src
+		SSoverlays.can_fire = TRUE
+
+/atom/proc/cut_overlay(image, priority)
+	if(!image)
+		return
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = LAZYLEN(cached_overlays)
+	var/init_p_len = LAZYLEN(cached_priority)  //starter pokemon
+	
+	LAZYREMOVE(cached_overlays, image)
+	if(priority)
+		LAZYREMOVE(cached_priority, image)
+
+	if((init_o_len != LAZYLEN(cached_priority)) || (init_p_len != LAZYLEN(cached_overlays)))
+		SSoverlays.processing[src] = src	//something was removed, queue it for compiling
+		SSoverlays.can_fire = TRUE
+
+/atom/proc/add_overlay(image, priority = FALSE)
+	if(!image)
+		return
+	LAZYINITLIST(our_overlays)	//always initialized after this point
+	LAZYINITLIST(priority_overlays)
+	var/list/cached_overlays = our_overlays	//sanic
+	var/list/cached_priority = priority_overlays
+	var/init_o_len = cached_overlays.len
+	var/init_p_len = cached_priority.len  //starter pokemon
+	var/need_compile
+
+	if(priority)
+		cached_priority |= image  //or in the image. Can we use [image] = image?
```

Comment:
I'm going to say that if a dev is adding the same overlay multiple times, its a bug or its intended and either we shouldn't hide it with |= or we are just getting in the way and they will have to do some hack.

remove the overhead, make the dev keep track of their shit.